update spec and fix code snippets

robert-zaremba opened the pull request:
## Description

closes: #9257

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

ryanchrypto commented on x/authz/spec/03_messages.md:
```diff
@@ -4,37 +4,40 @@ order: 3
 An allowed authorization can be removed with `MsgRevoke` message.
 
-+++ https://github.com/cosmos/cosmos-sdk/blob/c95de9c4177442dee4c69d96917efc955b5d19d9/proto/cosmos/authz/v1beta1/tx.proto#L53-L59
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L60-L64
```
```suggestion
+++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L58-L64
```

ryanchrypto replied:
This does not need to be changed. Maybe this is where you meant to respond with "ditto". Resolving the suggestion.

ryanchrypto reviewed (commented):
Looking good.

The code snippets are inconsistent. Some include the comment (those in the concepts and state documents), some do not (those in the messages document), and a few need to be fixed.

A lot of little suggestions. It would be nice if we were more consistent with how we referenced things throughout our documentation.

ryanchrypto commented:
It might also be nice to refer to the triple as `granter-authorization-grantee` rather than `(granter, grantee, Authorization)`, which would more closely reflect the RDF syntax as *subject-predicate-object*. Thoughts?

ryanchrypto commented on x/authz/spec/03_messages.md:
```diff
@@ -4,37 +4,40 @@ order: 3
 - Authorization Method doesn't exist (there is no defined handler in the app router to handle that Msg types)
 
-## Msg/Revoke
+If there is already a grant for `(granter, grantee, AuthorizationType)` triple, then the new grant will overwrite the previous one. To update or extend an existing grant a new grant with same `(granter, grantee, AuthorizationType)` should be created.
```
Also, see [comment](https://github.com/cosmos/cosmos-sdk/pull/9334#issuecomment-841708265).

ryanchrypto commented on x/authz/spec/02_state.md:
```diff
@@ -4,11 +4,12 @@ order: 2
+## Grant
 
-Grants are identified by combining granter address (the address bytes of the granter), grantee address (the address bytes of the grantee) and Msg type (its TypeURL). Hence we only allow one grant for the (granter, grantee, msg type) triple.
+Grants are identified by combining granter address (the address bytes of the granter), grantee address (the address bytes of the grantee) and Msg type (its type URL). Hence we only allow one grant for the (granter, grantee, msg type) triple.
```
Also, see [comment](https://github.com/cosmos/cosmos-sdk/pull/9334#issuecomment-841708265).

robert-zaremba replied:
`TypeURL` is not an attribute. We have a general method `sdk.MsgTypeURL`. Here I want to talk about an URL of a type in English.

ryanchrypto replied:
Ok. That makes sense. The triple reference in this paragraph should probably be consistent with the others.
```suggestion
Grants are identified by combining granter address (the address bytes of the granter), grantee address (the address bytes of the grantee) and Msg type (its type URL). Hence we only allow one grant for the (granter, grantee, Authorization) triple.
```

ryanchrypto commented on proto/cosmos/authz/v1beta1/tx.proto:
```diff
@@ -14,7 +14,8 @@ option (gogoproto.goproto_getters_all) = false;
 service Msg {
   // Grant grants the provided authorization to the grantee on the granter's
-  // account with the provided expiration time.
+  // account with the provided expiration time. If three is already a grant
+  // for given (grantee, granter, Authoriazation), then it will be overwritten.
```
Also, see [comment](https://github.com/cosmos/cosmos-sdk/pull/9334#issuecomment-841708265).

ryanchrypto replied:
```suggestion
  // account with the provided expiration time. If there is already a grant
  // for given the (grantee, granter, Authorization) triple, then the grant
  // will be overwritten.
```

robert-zaremba commented on x/authz/spec/03_messages.md:
```diff
@@ -4,37 +4,40 @@ order: 3
+In this section we describe the Msg Protobuf service for the authz module.
 
-## Msg/Grant
+## Grant
```
I wanted to discuss and describe service methods, not the request objects. That's why I'm not using `Msg` prefix.

ryanchrypto replied:
The title of this document is `Messages`. The next line of this section is "An authorization-grant is created using the `MsgGrant` message." Then we provide a code snippet for `MsgGrgant`. And then we further discuss the message, "This message is expected to fail if..." Switching to the method names as the titles for each section without changing the content or even mentioning the method is confusing.

robert-zaremba replied:
OK, I can rename the file and the header to `03_tx_service`.

robert-zaremba replied:
Each service has exactly one argument, so it makes sense to describe the argument (`MsgGrant`) in this case.

robert-zaremba commented on x/authz/spec/03_messages.md:
```diff
@@ -4,37 +4,40 @@ order: 3
 - provided `MsgTypeUrl` is empty.
 
-## Msg/Exec
+## Exec
```
ditto, let's continue discussion in one thread.

robert-zaremba commented on x/authz/spec/01_concepts.md:
```diff
@@ -9,10 +9,9 @@ order: 1
 
-
-+++ https://github.com/cosmos/cosmos-sdk/blob/c95de9c4177442dee4c69d96917efc955b5d19d9/x/authz/types/authorizations.go#L15-L24
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/x/authz/authorizations.go#L14-L25
```
I didn't want to include a comment here - because the description above covers it.

ryanchrypto replied:
Ok, that makes sense. This should be `L11-L25` then. The code snippet currently starts on `L14`, which is not the beginning of the declaration.

robert-zaremba commented on x/authz/spec/03_messages.md:
```diff
@@ -4,37 +4,40 @@ order: 3
 An authorization-grant is created using the `MsgGrant` message.
 
-+++ https://github.com/cosmos/cosmos-sdk/blob/c95de9c4177442dee4c69d96917efc955b5d19d9/proto/cosmos/authz/v1beta1/tx.proto#L27-L35
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L33-L37
```
ditto

ryanchrypto replied:
Not sure I understand "ditto" here. The declaration starts on line 32. This still needs to be fixed.

robert-zaremba replied:
yes, fixed it.

robert-zaremba commented:
> It might be nice to order the triple `(granter, Authorization, grantee)` rather than `(granter, grantee, Authorization)`, which would reflect the RDF order _subject-predicate-object_.

Interesting idea. I think we don't need this at this level. The order of this relation is hidden from the user, and it's a storage detail. We use `(granter, grantee, Authorization)` because we need a prefix scan based on `(granter, grantee)`.

ryanchrypto commented:
> Interesting idea. I think we don't need this at this level. The order of this relation is hidden from the user, and it's a storage detail. We use (granter, grantee, Authorization) because we need a prefix scan based on (granter, grantee).

Ok, just a thought. No need to change the order or how we refer to the triple.

A few items remaining:

1. There are two typos in the comment for the `Grant` method and we should probably be consistent with how we refer to the triple. See [comment](https://github.com/cosmos/cosmos-sdk/pull/9334#discussion_r633104903).
2. There is still one use of `msg type` rather than `Authorization` when referring to the triple that should probably be updated for consistency. See [comment](https://github.com/cosmos/cosmos-sdk/pull/9334#discussion_r633104160).
3. There are still a couple code snippets that do not start on the correct line (I understand the comment should not be included). See [comment](https://github.com/cosmos/cosmos-sdk/pull/9334#discussion_r633103196) and [comment](https://github.com/cosmos/cosmos-sdk/pull/9334#discussion_r632795181).
4. If we intend to discuss the methods rather than the messages, I think we need to update some of the content. It is not clear that we are describing the service methods. See [comment](https://github.com/cosmos/cosmos-sdk/pull/9334#discussion_r633104160).

robert-zaremba commented:
Thanks for review @ryanchrypto . I've renamed the 3_messages file to 3_services, updated snippets and used `Authorization` instead of Msg Type when we talk about grant.

ryanchristo reviewed (commented):
> I've renamed the 3_messages file to 3_services

In other module specifications, we describe the messages in a `messages.md` file and we do not have a `services.md` file. Do you think we should be breaking the pattern here?

The content for each section in the `services.md` file (renamed from `messages.md`) describes the messages. I am not sure why renaming each section to the service method names and then renaming the file to `services.md` are necessary changes if we are not updating the content to include explanations of the service methods.

Maybe it would make more sense if we added an explanation and a code snippet for each method? If we rename the file, we should probably use the singular `service.md` rather than `services.md`.

I have added more suggestions. I also think we could do a better job throughout our specs and documentation on using consistent naming conventions (i.e. when to use backticks, how to refer to Protobuf services and messages, etc.). It would be nice to have a styleguide that everyone could reference when writing documentation, but this is out of scope for this pull request.

robert-zaremba commented:
> In other module specifications, we describe the messages in a messages.md file and we do not have a services.md file. Do you think we should be breaking the pattern here?

This is what I'm proposing here. I think in recent discussions we want to promote the services oriented API.
@aaronc , @amaurym  - do you have an opinion here?

robert-zaremba commented on x/authz/spec/01_concepts.md:
```diff
@@ -21,22 +20,22 @@ Cosmos-SDK `x/authz` module comes with following authorization types
+`SendAuthorization` implements the `Authorization` interface for the `cosmos.bank.v1beta1.MsgSend` Msg. It takes a `SpendLimit` that specifies the maximum amount of tokens the grantee can spend, which is updated as the tokens are spent.
 
-+++ https://github.com/cosmos/cosmos-sdk/blob/c95de9c4177442dee4c69d96917efc955b5d19d9/proto/cosmos/authz/v1beta1/authz.proto#L12-L19
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/bank/v1beta1/authz.proto#L10-L19
```
Here I want to include a comment, because this is not well explained in the section above.

ryanchristo replied:
Ok. It might be nice to be consistent and to improve the explanation above rather than including the comment but up to you.

robert-zaremba commented on x/authz/spec/01_concepts.md:
```diff
@@ -21,22 +20,22 @@ Cosmos-SDK `x/authz` module comes with following authorization types
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/bank/v1beta1/authz.proto#L10-L19
 
-+++ https://github.com/cosmos/cosmos-sdk/blob/c95de9c4177442dee4c69d96917efc955b5d19d9/x/authz/types/send_authorization.go#L23-L45
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/x/bank/types/send_authorization.go#L25-L40
```
as above

robert-zaremba commented on x/authz/spec/03_services.md:
```diff
@@ -0,0 +1,46 @@
+
+# Tx Msg Service
+
+In this section we describe the Protobuf Msg service for the authz module.
```
not sure about it. `x/authz` emphasis package name, "authz module" is a plain English. Do you have a preference?

ryanchristo replied:
That makes sense. I do not have a strong preference. This would be a good example of something to explain in a styleguide for writing documentation/specifications (maybe added to SPEC-SPEC).

robert-zaremba commented on x/authz/spec/03_services.md:
```diff
@@ -0,0 +1,46 @@
+
+- authorization not implemented for the provided msg.
+- grantee don't have permission to run transaction.
+- if granted authorization is expired.
```
I think we want to talk in plain English here.

ryanchristo replied:
Ok. No need to change.

robert-zaremba commented on x/authz/spec/03_services.md:
```diff
@@ -0,0 +1,46 @@
+
+## Grant
+
+An authorization-grant is created using the `MsgGrant` message.
```
Correct, we should start describing the method. I have updated the section. Hope it sound better now. (same for comments below)

amaurym commented:
> This is what I'm proposing here. I think in recent discussions we want to promote the services oriented API.
@aaronc , @amaurym - do you have an opinion here?

we do have a SPEC-SPEC here: https://github.com/cosmos/cosmos-sdk/blob/master/docs/spec/SPEC-SPEC.md. IMO we should strictly follow the spec-spec for now.

We can start a separate issue/discussion about whether we want to update SPEC-SPEC. I agree we should promote the service-oriented design. However, I still think it's useful to separate query services and msg services into 2 files, so I wouldn't call it `03_services.md`. Maybe `03_query_service.md` and `04_msg_service.md`? Anyway, we should discuss this in a separate issue, and keep `03_messages.md` in this PR.

cyberbono3 reviewed (approved):
LGTM

ryanchristo commented:
> we do have a SPEC-SPEC here: https://github.com/cosmos/cosmos-sdk/blob/master/docs/spec/SPEC-SPEC.md. IMO we should strictly follow the spec-spec for now.
>
> We can start a separate issue/discussion about whether we want to update SPEC-SPEC. I agree we should promote the service-oriented design. However, I still think it's useful to separate query services and msg services into 2 files, so I wouldn't call it 03_services.md. Maybe 03_query_service.md and 04_msg_service.md? Anyway, we should discuss this in a separate issue, and keep 03_messages.md in this PR.

This makes sense to me. I think @robert-zaremba was moving in the direction of making valid improvements but we should discuss updating the spec-spec document before changing the structure here.

aaronc commented:
> > we do have a SPEC-SPEC here: https://github.com/cosmos/cosmos-sdk/blob/master/docs/spec/SPEC-SPEC.md. IMO we should strictly follow the spec-spec for now.
> > We can start a separate issue/discussion about whether we want to update SPEC-SPEC. I agree we should promote the service-oriented design. However, I still think it's useful to separate query services and msg services into 2 files, so I wouldn't call it 03_services.md. Maybe 03_query_service.md and 04_msg_service.md? Anyway, we should discuss this in a separate issue, and keep 03_messages.md in this PR.
> 
> This makes sense to me. I think @robert-zaremba was moving in the direction of making valid improvements but we should discuss updating the spec-spec document before changing the structure here.

I think this approach that @ryanchristo is suggesting makes sense

robert-zaremba commented:
We discussed it during the stand up and arrived to the same conclusion

* we will use _messages_ oriented spec here
* I created a task to update the spec and authz module spec: https://github.com/cosmos/cosmos-sdk/issues/9352  . Other modules can  follow after that.

ryanchristo commented on x/authz/spec/03_messages.md:
```diff
@@ -6,38 +6,42 @@ order: 3
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.43.0-beta1/proto/cosmos/authz/v1beta1/tx.proto#L32-L37
 
-This message is expected to fail if:
+The message handling should to fail if:
```
```suggestion
The message handling should fail if:
```

ryanchristo commented on x/authz/spec/03_messages.md:
```diff
@@ -6,38 +6,42 @@ order: 3
+## MsgRevoke
 
-An allowed authorization can be removed with `MsgRevoke` message.
+A grant can be removed with `MsgRevoke` message.
```
```suggestion
A grant can be removed with the `MsgRevoke` message.
```

ryanchristo commented on x/authz/spec/03_messages.md:
```diff
@@ -6,38 +6,42 @@ order: 3
 - provided `MsgTypeUrl` is empty.
 
-## Msg/Exec
+NOTE: `MsgExec` message removes a grant if the grant exhausted.
```
```suggestion
NOTE: The `MsgExec` message removes a grant if the grant has expired.
```

ryanchristo commented on x/authz/spec/03_messages.md:
```diff
@@ -6,38 +6,42 @@ order: 3
+## MsgExec
 
-+++ https://github.com/cosmos/cosmos-sdk/blob/c95de9c4177442dee4c69d96917efc955b5d19d9/proto/cosmos/authz/v1beta1/tx.proto#L42-L48
+When a grantee wants to execute transaction on behalf of a granter, it must send `MsgExec`.
```
```suggestion
When a grantee wants to execute a transaction on behalf of a granter, they must send `MsgExec`.
```

ryanchristo commented on x/authz/spec/03_messages.md:
```diff
@@ -6,38 +6,42 @@ order: 3
 
-- authorization not implemented for the provided msg.
-- grantee don't have permission to run transaction.
+The method should fail if:
```
```suggestion
The message handling should fail if:
```

ryanchristo reviewed (approved):
Looks good to me! A few nitpicks but I approve.

aleem1314 reviewed (approved):
lgtm

robert-zaremba commented:
The GitHub CI strikes again.
